File: FluentFTP.CSharpExamples/GetListingWithLinks.cs
using System;
using System.Net;
using System.Threading;
using System.Threading.Tasks;
using FluentFTP;

namespace Examples {
	internal static class GetListingWithLinksExample {

		public static void GetListing() {
			using (var conn = new FtpClient("127.0.0.1", "ftptest", "ftptest")) {
				conn.Connect();

				// get a listing of the files with their modified date & size
				foreach (var item in conn.GetListing(conn.GetWorkingDirectory(),
				FtpListOption.Modify | FtpListOption.Size)) {
					switch (item.Type) {
						case FtpObjectType.Directory:
							break;

						case FtpObjectType.File:
							break;

						case FtpObjectType.Link:

							// dereference symbolic links
							if (item.LinkTarget != null) {
								// see the DereferenceLink() example
								// for more details about resolving links.
								item.LinkObject = conn.DereferenceLink(item);

								if (item.LinkObject != null) {
									// switch (item.LinkObject.Type)...
								}
							}

							break;
					}
				}

				// get a listing of the files with their modified date & size, and automatically dereference links
				foreach (var item in conn.GetListing(conn.GetWorkingDirectory(),
					FtpListOption.Modify | FtpListOption.Size | FtpListOption.DerefLinks)) {
					switch (item.Type) {
						case FtpObjectType.Directory:
							break;

						case FtpObjectType.File:
							break;

						case FtpObjectType.Link:
							if (item.LinkObject != null) {
								// switch (item.LinkObject.Type)...
							}

							break;
					}
				}
			}
		}

		public static async Task GetListingAsync() {
			var token = new CancellationToken();
			using (var conn = new FtpClient("127.0.0.1", "ftptest", "ftptest")) {
				await conn.ConnectAsync(token);

				// get a listing of the files with their modified date & size
				foreach (var item in await conn.GetListingAsync(conn.GetWorkingDirectory(),
				FtpListOption.Modify | FtpListOption.Size, token)) {
					switch (item.Type) {
						case FtpObjectType.Directory:
							break;

						case FtpObjectType.File:
							break;

						case FtpObjectType.Link:

							// dereference symbolic links
							if (item.LinkTarget != null) {
								// see the DereferenceLink() example
								// for more details about resolving links.
								item.LinkObject = conn.DereferenceLink(item);

								if (item.LinkObject != null) {
									// switch (item.LinkObject.Type)...
								}
							}

							break;
					}
				}

				// get a listing of the files with their modified date & size, and automatically dereference links
				foreach (var item in await conn.GetListingAsync(conn.GetWorkingDirectory(),
					FtpListOption.Modify | FtpListOption.Size | FtpListOption.DerefLinks, token)) {
					switch (item.Type) {
						case FtpObjectType.Directory:
							break;

						case FtpObjectType.File:
							break;

						case FtpObjectType.Link:
							if (item.LinkObject != null) {
								// switch (item.LinkObject.Type)...
							}

							break;
					}
				}
			}
		}

	}
}

File: FluentFTP.VBExamples/GetListingWithLinks.vb
Imports System
Imports System.Net
Imports System.Threading
Imports System.Threading.Tasks
Imports FluentFTP

Namespace Examples
	Friend Module GetListingWithLinksExample
		Sub GetListing()
			Using conn = New FtpClient("127.0.0.1", "ftptest", "ftptest")
				conn.Connect()

				' get a listing of the files with their modified date & size
				For Each item In conn.GetListing(conn.GetWorkingDirectory(), FtpListOption.Modify Or FtpListOption.Size)

					Select Case item.Type

						Case FtpObjectType.Directory
						Case FtpObjectType.File
						Case FtpObjectType.Link

							' manually dereference symbolic links
							If item.LinkTarget IsNot Nothing Then
								item.LinkObject = conn.DereferenceLink(item)

								If item.LinkObject IsNot Nothing Then
								End If
							End If

					End Select
				Next

				' get a listing of the files with their modified date & size, and automatically dereference links
				For Each item In conn.GetListing(conn.GetWorkingDirectory(), FtpListOption.Modify Or FtpListOption.Size Or FtpListOption.DerefLinks)

					' do something
					Select Case item.Type

						Case FtpObjectType.Directory
						Case FtpObjectType.File
						Case FtpObjectType.Link

					End Select
				Next
			End Using
		End Sub

		Async Function GetListingAsync() As Task
			Dim token = New CancellationToken()

			Using conn = New FtpClient("127.0.0.1", "ftptest", "ftptest")
				Await conn.ConnectAsync(token)

				' get a listing of the files with their modified date & size
				For Each item In Await conn.GetListingAsync(conn.GetWorkingDirectory(), FtpListOption.Modify Or FtpListOption.Size, token)

					Select Case item.Type

						Case FtpObjectType.Directory
						Case FtpObjectType.File
						Case FtpObjectType.Link

							' manually dereference symbolic links
							If item.LinkTarget IsNot Nothing Then
								item.LinkObject = conn.DereferenceLink(item)

								If item.LinkObject IsNot Nothing Then
								End If
							End If

					End Select
				Next

				' get a listing of the files with their modified date & size, and automatically dereference links
				For Each item In Await conn.GetListingAsync(conn.GetWorkingDirectory(), FtpListOption.Modify Or FtpListOption.Size Or FtpListOption.DerefLinks, token)

					' do something
					Select Case item.Type

						Case FtpObjectType.Directory
						Case FtpObjectType.File
						Case FtpObjectType.Link

					End Select
				Next
			End Using
		End Function
	End Module
End Namespace
